File: task1.py
from pyspark import SparkContext
import sys
import time
from collections import defaultdict

input_file = sys.argv[3]

sc = SparkContext('local[*]', 'task1')
sc.setLogLevel("OFF")
start = time.time()
readed_file = sc.textFile(input_file)
title = readed_file.first()
baskets = readed_file.filter(lambda x: x != title)

case = int(sys.argv[1])
nn = int(sys.argv[2])
output_file = sys.argv[4]


def makePair(x):
    dicc = {}
    k = 1
    support = nn / numberOfPartitions
    baskets_1 = [i[1] for i in list(x)]
    temp = []
    final_temp = []
    for subbasket in baskets_1:
        for j in subbasket:
            if j not in dicc:
                dicc[j] = 1
            else:
                dicc[j] += 1

    for i, j in dicc.items():
        if j >= support:
            temp.append({i})
            final_temp.append((i, 1))

    while (len(temp) > 0):
        k += 1

        if k == 2:
            temp_1 = []
            for i, j in enumerate(temp):
                for x, y in enumerate(temp):
                    if i < x:
                        temp_1.append(j.union(y))
            temp = []
            for pair in temp_1:
                support_lite = 0
                for basket in baskets_1:
                    if pair.issubset(basket):
                        support_lite += 1
                if support_lite >= support:
                    temp.append(pair)

            if len(temp) > 0:
                for ii in temp:
                    final_temp.append((tuple(sorted(ii)), 1))

        else:
            temp_2 = []
            for q, j in enumerate(temp):
                for x, y in enumerate(temp):
                    if q < x:
                        temp_3 = sorted(list(j))
                        temp_4 = sorted(list(y))
                        if (temp_3[:k - 2] == temp_4[:k - 2]):
                            to_add = set(j) | set(y)
                            temp_2.append(to_add)
            temp = []

            for pair_1 in temp_2:
                support_lite_1 = 0
                for basket in baskets_1:
                    if pair_1.issubset(basket):
                        support_lite_1 += 1
                if support_lite_1 >= support:
                    temp.append(pair_1)

            if len(temp) > 0:
                for jj in temp:
                    final_temp.append((tuple(sorted(jj)), 1))

    return final_temp


def final_touch(container, candidates):
    counting_dic = defaultdict(int)
    baskets = [i[1] for i in container]
    for i in candidates:

        if type(i) == str:
            for b in baskets:
                if i in b:
                    counting_dic[i] += 1
        else:
            for b in baskets:
                if set(i).issubset(set(b)):
                    counting_dic[i] += 1
    return counting_dic.items()


if case == 1:
    baskets = baskets.map(lambda x: tuple(x.split(','))).distinct().map(lambda x: (x[0], [x[1]])).reduceByKey(
        lambda a, b: a + b).persist()
elif case == 2:
    baskets = baskets.map(lambda x: tuple(x.split(','))).distinct().map(lambda x: (x[1], [x[0]])).reduceByKey(
        lambda a, b: a + b).persist()

numberOfPartitions = baskets.getNumPartitions()

stepone = baskets.mapPartitions(lambda x: makePair(x)).groupByKey().map(lambda x: x[0]).collect()
output_candidate = defaultdict(list)
for item in stepone:
    if type(item) == str:
        output_candidate[1].append(item)
    else:
        output_candidate[len(item)].append(item)

file = "Candidates:\n"
for key in output_candidate.keys():
    output_candidate[key].sort()
    if key == 1:
        temp = []
        for i in output_candidate[key]:
            temp.append("('" + i + "')")
        file += ','.join(temp) + "\n\n"
    else:
        final = []
        for j in output_candidate[key]:
            temp1 = []
            for ii in j:
                temp1.append("'" + ii + "'")
            final.append("(" + ', '.join(temp1) + ")")
        file += ','.join(final) + "\n\n"

steptwo = baskets.mapPartitions(lambda x: final_touch(x, stepone)).reduceByKey(lambda x, y: x + y).filter(
    lambda x: x[1] >= nn).map(lambda x: x[0]).collect()
output_frequent = defaultdict(list)
for item in steptwo:
    if type(item) == str:
        output_frequent[1].append(item)
    else:
        output_frequent[len(item)].append(item)

file += "Frequent Itemsets:\n"
for key in output_frequent.keys():
    output_frequent[key].sort()
    if key == 1:
        temp = []
        for i in output_frequent[key]:
            temp.append("('" + i + "')")
        file += ','.join(temp) + "\n\n"
    else:
        final = []
        for j in output_frequent[key]:
            temp1 = []
            for ii in j:
                temp1.append("'" + ii + "'")
            final.append("(" + ', '.join(temp1) + ")")
        file += ','.join(final) + "\n\n"

with open(output_file, 'w') as file1:
    file1.write(file)

print("Duration:", time.time() - start)

File: task1.scala
import java.io._

import org.apache.log4j.{Level, Logger}
import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

import scala.util.control.Breaks._
import scala.collection.mutable.ArrayBuffer

object task1 {


  def main(args: Array[String]): Unit = {
    val start = System.currentTimeMillis()
    val conf = new SparkConf().setAppName("task21").setMaster("local[*]")
    var sc = new SparkContext(conf)
    Logger.getLogger("org").setLevel(Level.FATAL)
    val s1=System.nanoTime()
    val case_ = args(0).toInt
    val support_1 = args(1).toInt
    val input1 = args(2)
    val output1 = args(3)

    var data = sc.textFile(input1)
    var header = data.first()
    data = data.filter(x => x != header)




    def getinformat[A : Ordering]
    (item: Seq[Iterable[A]]) = item.sorted

    var baskets_1: RDD[Set[String]] = null
    if (case_ == 1) {
      baskets_1 = data.map(x => x.split(",")).map(x => (x(0), x(1))).distinct().groupByKey().map(x => x._2.toSet)
      //baskets_1.foreach(println)

    } else {
      baskets_1 = data.map(x => x.split(",")).map(x => (x(1), x(0))).distinct().groupByKey().map(x => x._2.toSet)
    }

    var n = data.getNumPartitions

    def makepair(basket: Iterator[Set[String]], n: Int): Iterator[Set[String]] = {
      var pool = basket.toList
      var support_2 = (support_1 / n)
      var temp = pool.flatten.groupBy(identity).mapValues(_.map(_ => 1).sum).filter(x => x._2 >= support_2).map(x => x._1).toSet


      var final_temp = Set.empty[Set[String]]
      for (i <- temp) {
        final_temp = final_temp + Set(i)
      }

      var total = final_temp.size
      var temp_1 = Set.empty[String]
      var k = 2
      while (temp.size > 0) {
        if (k == 2) {
          var pairs = temp.subsets(2)
          for (i <- pairs) {
            breakable {
              var count = 0
              for (j <- pool) {
                if (i.subsetOf(j)) {
                  count = count + 1
                  if (count >= support_2) {
                    temp_1 = temp_1 ++ i
                    final_temp = final_temp + i
                    break
                  }
                }
              }
            }
          }
          k = k + 1
          temp = temp_1
          temp_1 = Set.empty[String]
        }
        else {
          //println("this is temp", temp, k)
          var tempu = temp.subsets(k)
          for (i <- tempu) {
            breakable {
              var count = 0
              for (j <- pool) {
                if (i.subsetOf(j)) {
                  count = count + 1
                  if (count >= support_2) {
                    temp_1 = temp_1 ++ i
                    final_temp = final_temp + i
                    break
                  }
                }
              }
            }
          }
          k = k + 1
          temp = temp_1
          //println("this is temp_1 before", temp_1)
          temp_1 = Set.empty[String]
          //println("this is temp_1 after", temp_1)

        }
      }

      return final_temp.iterator
    }

    var phase1 = baskets_1.mapPartitions(x => makepair(x, n)).distinct().map(x => (x, 1)).reduceByKey((x, y) => 1).map(_._1).collect()

    def finaltouch(basket: Iterator[Set[String]], ph: Array[Set[String]]): Iterator[(Set[String], Int)] = {
      var pool1 = basket.toList
      var outcome = List[(Set[String], Int)]()
      for (ii <- pool1)
      {
        for (jj <- ph)
        {
          if (jj.forall(ii.contains))
          {
            outcome = Tuple2(jj, 1) :: outcome
          }
        }
      }
      return outcome.iterator
    }: Iterator[(Set[String], Int)]

    var phase_2 = baskets_1.mapPartitions(x => finaltouch(x, phase1)).reduceByKey((a,b) => a+b).filter(_._2 >= support_1)
    var formatted_phase_2 = phase_2.map(x =>x._1).map(x => (x,x.size)).collect()

    var max = 0
    for (i <- formatted_phase_2)
    {
      if (i._2 > max)
      {
        max = i._2
      }
    }
    var single = getinformat(formatted_phase_2.filter(x => x._2 == 1).map(x => x._1))
    var output = "Candidates:\n"
    for (i<- single)
    {
      if (i==single.last){

        output += "('"+ i.toList.head + "')\n\n"
      }
      else {
        output += "('"+ i.toList.head + "')" + ","
      }
    }

    for (j <- 2 to max)
    {

      var size_ : Int= formatted_phase_2.size
      var internal_sorted=  ArrayBuffer[(List[String],Int)]()
      for(i_internal<- formatted_phase_2){
        internal_sorted .+=( Tuple2(i_internal._1.toList.sorted,i_internal._2))
      }
      var thisis = getinformat(internal_sorted.filter(x => x._2 == j).map(x => x._1))


      for (jj <- thisis) {
        var jjs = jj.toList.sorted
        output += "("
        for (zz <- jjs) {
          output += "'" + zz + "'" + "," + " "
        }
        output = output.slice(0, output.length - 2) + "),"
      }
      output = output.patch(output.lastIndexOf(','), "", 1) + "\n\n"

    }

    output += "Frequent Itemsets:\n"
    for (i<- single)
    {
      if (i==single.last){

        output += "('"+ i.toList.head + "')\n\n"
      }
      else {
        output += "('"+ i.toList.head + "')" + ","
      }
    }

    for (j <- 2 to max)
    {

      var size_ : Int= formatted_phase_2.size
      var internal_sorted=  ArrayBuffer[(List[String],Int)]()
      for(i_internal<- formatted_phase_2){
        internal_sorted .+=( Tuple2(i_internal._1.toList.sorted,i_internal._2))
      }
      var thisis = getinformat(internal_sorted.filter(x => x._2 == j).map(x => x._1))


      for (jj <- thisis) {
        var jjs = jj.toList.sorted
        output += "("
        for (zz <- jjs) {
          output += "'" + zz + "'" + "," + " "
        }
        output = output.slice(0, output.length - 2) + "),"
      }
      output = output.patch(output.lastIndexOf(','), "", 1) + "\n\n"

    }

    val printWriter = new PrintWriter(new File(output1))
    printWriter.write(output)
    printWriter.close()

    val s2=System.nanoTime()
    val t2=(s2-s1)/1000000000.0
    println("Duration:"+t2.toString())

  }

}

